src/combinatorics/simple_graph/prod.lean

```diff
@@ -173,30 +173,35 @@ by { haveI := (nonempty_prod.1 h.nonempty).1, haveI := (nonempty_prod.1 h.nonemp
 @[simp] lemma box_prod_connected : (G □ H).connected ↔ G.connected ∧ H.connected :=
 ⟨λ h, ⟨h.of_box_prod_left, h.of_box_prod_right⟩, λ h, h.1.box_prod h.2⟩
 
-instance [decidable_eq α] [decidable_eq β] (x : α × β)
+instance box_prod_fintype_neighbor_set (x : α × β)
   [fintype (G.neighbor_set x.1)] [fintype (H.neighbor_set x.2)] :
   fintype ((G □ H).neighbor_set x) :=
-begin
-  rw box_prod_neighbor_set,
-  apply_instance,
-end
+fintype.of_equiv
+  ((G.neighbor_finset x.1 ×ˢ {x.2}).disj_union ({x.1} ×ˢ H.neighbor_finset x.2)
+      $ λ y hG hH, begin
+        simp_rw [finset.mem_product, finset.mem_singleton,
+          mem_neighbor_finset] at hG hH,
+        exact hG.1.ne hH.1.symm,
+      end)
+  ((equiv.refl _).subtype_equiv $ λ y, begin
+    simp_rw [finset.mem_disj_union, finset.mem_product, finset.mem_singleton,
+          mem_neighbor_finset, mem_neighbor_set, equiv.refl_apply, box_prod_adj],
+    simp only [eq_comm, and_comm],
+  end)
+
+lemma box_prod_neighbor_finset (x : α × β)
+  [fintype (G.neighbor_set x.1)] [fintype (H.neighbor_set x.2)] :
+  (G □ H).neighbor_finset x =
+    (G.neighbor_finset x.1 ×ˢ {x.2}).disj_union ({x.1} ×ˢ H.neighbor_finset x.2)
+      (box_prod_fintype_neighbor_set._proof_1 x) :=
+(finset.univ.map_map _ $ function.embedding.subtype _).trans finset.attach_map_val
 
 lemma box_prod_degree (x : α × β)
-  [fintype (G.neighbor_set x.1)] [fintype (H.neighbor_set x.2)]
-  [fintype ((G □ H).neighbor_set x)] :
+  [fintype (G.neighbor_set x.1)] [fintype (H.neighbor_set x.2)] :
```

Comment:
Do we want to drop the `[fintype ((G □ H).neighbor_set x)]` argument? I thought this makes the lemma less general in that you can't always `rw` with it, since you might have a `fintype` instance that arose by other means.

Reply:
You're right in principle, but in general we seem to assume that there's only one canonical fintype instance in the first place; for instance, [`fintype.card_prod`](https://leanprover-community.github.io/mathlib_docs/data/fintype/basic.html#fintype.card_prod) does not have such an argument.

Reply:
My feeling is that this is an error in fintype.card_prod rather than a pattern we should repeat - could we keep the (seemingly) unnecessary argument for the sake of generality? We are only gaining usability, at basically no cost

Reply:
Unfortunately it makes the proof rather ugly, but I can make the change.